### ISSUE TITLE
Download permissions broken. Unable to grant access or see permissions already granted in order.

### ISSUE THREAD

**Issue by seb86** [woocommerce/woocommerce#20412]
**Describe the bug**
I just had a customer order and for the first time, they were not granted access to the downloadable file applied to the product. I tried to manually add it myself but it won't let me.

It gives the following message in an alert box instead.

> Could not grant access - the user may already have permission for this file or the billing email is not set. Ensure the billing email is set, and the order has been saved.

Well, the order is completed, the customer has provided an email address under billing when making the order and does not have access to the file.

**What happened?**

**Expected behavior**
Product should automatically provide access to the downloadable file to the order and if not I should be able to add it manually.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
`
### WordPress Environment ###

Home URL: https://autoloadnextpost.com
Site URL: https://autoloadnextpost.com
WC Version: 3.4.2
Log Directory Writable: ✔
WP Version: 4.9.6
WP Multisite: –
WP Memory Limit: 768 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_GB
External object cache: –

### Server Environment ###

Server Info: LiteSpeed
PHP Version: 7.0.30 - We recommend using PHP version 7.2 or above for greater performance and security. How to update your PHP version
PHP Post Max Size: 128 MB
PHP Time Limit: 300
PHP Max Input Vars: 10000
cURL Version: 7.54.0
OpenSSL/1.0.2k

SUHOSIN Installed: –
MySQL Version: 5.6.23-cll-lve
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.4.2
WC Database Prefix: alnpwp_
MaxMind GeoIP Database: ✔
Total Database Size: 49.51MB
Database Data Size: 40.91MB
Database Index Size: 8.60MB
alnpwp_woocommerce_sessions: Data: 0.05MB + Index: 0.02MB
alnpwp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
alnpwp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
alnpwp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
alnpwp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
alnpwp_woocommerce_order_itemmeta: Data: 0.05MB + Index: 0.03MB
alnpwp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
alnpwp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
alnpwp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
alnpwp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB
alnpwp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
alnpwp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
alnpwp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
alnpwp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
alnpwp_commentmeta: Data: 0.02MB + Index: 0.03MB
alnpwp_comments: Data: 0.05MB + Index: 0.09MB
alnpwp_imagify_files: Data: 0.02MB + Index: 0.08MB
alnpwp_imagify_folders: Data: 0.02MB + Index: 0.03MB
alnpwp_links: Data: 0.02MB + Index: 0.02MB
alnpwp_nep_native_emoji: Data: 0.02MB + Index: 0.00MB
alnpwp_options: Data: 3.14MB + Index: 1.02MB
alnpwp_postmeta: Data: 11.41MB + Index: 0.67MB
alnpwp_posts: Data: 2.17MB + Index: 0.39MB
alnpwp_swp_cf: Data: 0.02MB + Index: 0.05MB
alnpwp_swp_index: Data: 1.50MB + Index: 0.39MB
alnpwp_swp_log: Data: 0.02MB + Index: 0.05MB
alnpwp_swp_tax: Data: 0.02MB + Index: 0.05MB
alnpwp_swp_terms: Data: 0.16MB + Index: 0.23MB
alnpwp_termmeta: Data: 0.02MB + Index: 0.03MB
alnpwp_terms: Data: 0.08MB + Index: 0.09MB
alnpwp_term_relationships: Data: 0.06MB + Index: 0.05MB
alnpwp_term_taxonomy: Data: 0.09MB + Index: 0.13MB
alnpwp_usermeta: Data: 0.08MB + Index: 0.06MB
alnpwp_users: Data: 0.02MB + Index: 0.05MB
alnpwp_wc_download_log: Data: 0.02MB + Index: 0.03MB
alnpwp_wc_webhooks: Data: 0.02MB + Index: 0.02MB
alnpwp_woocommerce_software_activations: Data: 0.02MB + Index: 0.00MB
alnpwp_woocommerce_software_licenses: Data: 0.02MB + Index: 0.00MB
alnpwp_wpforms_entries: Data: 0.06MB + Index: 0.02MB
alnpwp_wpforms_entry_fields: Data: 0.02MB + Index: 0.05MB
alnpwp_wpforms_entry_meta: Data: 0.02MB + Index: 0.02MB
alnpwp_wpgdprc_consents: Data: 0.02MB + Index: 0.00MB
alnpwp_wptc_activity_log: Data: 0.16MB + Index: 0.08MB
alnpwp_wptc_auto_backup_record: Data: 0.02MB + Index: 0.00MB
alnpwp_wptc_backups: Data: 0.02MB + Index: 0.02MB
alnpwp_wptc_backup_names: Data: 0.02MB + Index: 0.00MB
alnpwp_wptc_current_process: Data: 0.02MB + Index: 0.02MB
alnpwp_wptc_debug_log: Data: 0.05MB + Index: 0.00MB
alnpwp_wptc_excluded_files: Data: 0.02MB + Index: 0.00MB
alnpwp_wptc_excluded_tables: Data: 0.02MB + Index: 0.02MB
alnpwp_wptc_included_files: Data: 0.02MB + Index: 0.00MB
alnpwp_wptc_included_tables: Data: 0.02MB + Index: 0.00MB
alnpwp_wptc_options: Data: 0.03MB + Index: 0.00MB
alnpwp_wptc_processed_files: Data: 11.52MB + Index: 4.39MB
alnpwp_wptc_processed_iterator: Data: 0.02MB + Index: 0.00MB
alnpwp_wptc_processed_restored_files: Data: 0.02MB + Index: 0.00MB
alnpwp_yoast_seo_links: Data: 0.05MB + Index: 0.02MB
alnpwp_yoast_seo_meta: Data: 9.42MB + Index: 0.00MB

### Post Type Counts ###

amn_mi-lite: 2
attachment: 73
carts: 4
easy-pricing-table: 4
give_forms: 2
hollerbox: 1
mc4wp-form: 2
nav_menu_item: 51
page: 46
post: 17
product: 10
product_variation: 10
scheduled-action: 9
schema: 2
secupress_log_action: 1000
secupress_log_err404: 1000
shop_coupon: 3
shop_order: 26
shop_order_refund: 3
theme: 15
user_request: 1
wp_block: 1
wpforms: 3

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (73) ###

WP Time Capsule: by Revmakx – 1.17.2
Akismet Anti-Spam: by Automattic – 4.0.7
WP Offload S3: by Delicious Brains – 1.6.2
Amazon Web Services: by Delicious Brains – 1.0.5
AMP: by WordPress.com VIP
XWP
Google
and contributors – 0.7.1

Attachment Slugs for WordPress: by Sébastien Dumont – 0.0.2
Cart REST API for WooCommerce: by Sébastien Dumont – 1.0.3 – Not tested with the active version of WooCommerce
Code Prettify: by Kaspars Dambis – 1.3.4
Disable Cart Fragments: by LittleBizzy – 1.0.9 – Not tested with the active version of WooCommerce
Disable Feeds: by Samir Shah
BBpress support by Matthew Sigley – 1.4.4

Duplicate Post: by Enrico Battocchi – 3.2.2
Easy Plugin Stats: by Nick Diego – 1.0.0
Easy Pricing Tables Premium: Business: by Fatcat Apps – 2.3.10
Enable Media Replace: by ShortPixel – 3.2.5
Facebook for WooCommerce: by Facebook – 1.8.7 – Not tested with the active version of WooCommerce
Glue for Yoast SEO & AMP: by Joost de Valk – 0.4.3
Google Analytics for WordPress by MonsterInsights: by MonsterInsights – 7.0.6
Google Authenticator - Per User Prompt: by Ian Dunn – 0.7
Google Authenticator: by Henrik Schack – 0.48
Heartbeat Control: by Jeff Matson – 1.2.3
Holler Box: by Scott Bolinger – 1.4.1
Holler Box Pro: by Scott Bolinger – 0.7.0
Holler Box FOMO Notifications: by Scott Bolinger – 0.6.1 – Not tested with the active version of WooCommerce
Imagify: by WP Media – 1.7.1.3
Instant Images: by Darren Cooney – 3.1
Jetpack by WordPress.com: by Automattic – 6.2
Jilt for WooCommerce: by Jilt – 1.4.5
Login Designer: by Rich Tabor from ThatPluginCompany – 1.1.8
MailChimp for WordPress: by ibericode – 4.2.2
MailChimp Activity: by ibericode – 1.0.5
Native Emoji: by Daniel Brandenburg – 3.0
Nav Menu Roles: by Kathy Darling – 1.9.2
No Self Pings: by Michael D. Adams – 1.1.2
Plugins Site Menu Link: by Sébastien Dumont – 1.1.0
Public Post Preview: by Dominik Schilling – 2.6.0
Remove Google Analytics (MonsterInsights) comments: by Mitch – 1.0.1
Remove Query Strings: by LittleBizzy – 1.3.0
Schema: by Hesham – 1.6.9.8.2
SearchWP Diagnostics: by Jonathan Christopher – 1.4.1
SearchWP WooCommerce Integration: by SearchWP
LLC – 1.1.20 – Not tested with the active version of WooCommerce

SearchWP: by SearchWP
LLC – 2.9.13

SecuPress Free — WordPress Security: by SecuPress – 1.4.4
Shortcodes in Sidebar: by pankajanupam – 3.0
Simply Show Hooks: by Stuart O'Brien
cxThemes – 1.2.1

Page Builder by SiteOrigin: by SiteOrigin – 2.6.8
SiteOrigin Widgets Bundle: by SiteOrigin – 1.11.8
Theme Designer: by Justin Tadlock – 1.0.0
Cookie Consent: by Catapult_Themes – 2.3.10
User Switching: by John Blackbourn & contributors – 1.3.1
Users by Date Registered: by Sébastien Dumont – 1.0.3
What The File: by Never5 – 1.5.4
Widgets for SiteOrigin: by wpinked – 1.3.6
WooCommerce Cart Reports: by WooCommerce – 1.1.20 – Not tested with the active version of WooCommerce
WooCommerce Coupon Campaigns & Tracking: by WooCommerce – 1.1.2
WooCommerce Stripe Gateway: by WooCommerce – 4.1.6
WooCommerce Google Analytics Integration: by WooCommerce – 1.4.4 – Not tested with the active version of WooCommerce
WooCommerce Name Your Price: by Kathy Darling – 2.8.1 – 2.8.2 is available – Not tested with the active version of WooCommerce
WooCommerce Order Status Control: by SkyVerge – 1.9.0 – 1.9.1 is available – Not tested with the active version of WooCommerce
WooCommerce Pre-Orders: by WooCommerce – 1.5.9
WooCommerce Product Add-ons: by WooCommerce – 2.9.6
WooCommerce Services: by Automattic – 1.14.1 – Not tested with the active version of WooCommerce
WooCommerce Software Add-On: by WooCommerce – 1.7.4 – 1.7.5 is available – Not tested with the active version of WooCommerce
WooCommerce Subscriptions: by Prospress Inc. – 2.2.20 – 2.2.21 is available – Not tested with the active version of WooCommerce
WooCommerce URL Coupons: by SkyVerge – 2.6.1 – 2.6.3 is available – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.4.2
Yoast SEO Premium: by Team Yoast – 7.6
WP GDPR Compliance: by Van Ons – 1.3.9
WP Migrate DB Pro Media Files: by Delicious Brains – 1.4.9
WP Migrate DB Pro: by Delicious Brains – 1.8.1
WP Rocket | External JS Hosts: by WP Rocket Support Team –
WP Rocket: by WP Media – 3.0.5.1
WPForms: by WPForms – 1.4.7
Yoast SEO: WooCommerce: by Team Yoast – 7.6 – Not tested with the active version of WooCommerce

### Settings ###

API Enabled: –
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

### WC Pages ###

Shop base: #30 - /products/
Basket: #31 - /cart/
Checkout: #32 - /checkout/
My account: #33 - /my-account/
Terms and conditions: #1595 - /terms-and-conditions/

### Theme ###

Name: Auto Load Next Post Theme
Version: 1.0
Author URL: https://sebastiendumont.com
Child Theme: ✔
Parent Theme Name: Twenty Seventeen
Parent Theme Version: 1.6
Parent Theme Author URL: https://wordpress.org/
WooCommerce Support: ✔

### Templates ###

Overrides: auto-load-next-post-theme/woocommerce/cart/cart-empty.php
auto-load-next-post-theme/woocommerce/myaccount/dashboard.php
auto-load-next-post-theme/woocommerce/myaccount/downloads.php
auto-load-next-post-theme/woocommerce/single-product/meta.php

### subscriptions ###

WCS_DEBUG: ✔
						No
Subscriptions Mode: ✔
						Live
Subscriptions Template Theme Overrides: –
Report Cache Enabled: ✔
						Yes
Cache Update Failures: ✔
						0 failure

### Jilt ###

Plugin Version: 1.4.5
API Version: v2
API Authentication Method: OAuth
API Connected: ✔
Linked to Jilt: ✔
Enabled: ✔
`
</details>

**Comment by mikejolley** [comment]
What is the type of the download_id field in your database? Specifically its length.

**Comment by seb86** [comment]
Table: _alnpwp_woocommerce_downloadable_product_permissions_

The download_id field is `varchar(36)`

**Comment by seb86** [comment]
I can confirm that the user has not been granted permission according to the table in the database.

**Comment by seb86** [comment]
Why does the billing address have to be required when Ajax runs? https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-ajax.php#L762|L764

**Comment by seb86** [comment]
I looked at a previous order from another customer last year for the same product which I know for a fact had been given permission and I can see it permission has been given in table in the database but the order will not show it so there is clearly a bug.

**Comment by seb86** [comment]
I can search for the downloadable item but any previous permissions are not shown and any new ones granted will not add.

**Comment by seb86** [comment]
On a stage site I have switched to the Storefront theme and disabled all plugins except for the following:

- WooCommerce
- WooCommerce Services
- WooCommerce Stripe Gateway
- WP Migrate DB Pro
- WP Migrate DB Pro Media Files
- WP Offload S3

Still getting the same results. I even tried a manual order and still unable to grant permission.

**Comment by seb86** [comment]
@mikejolley Could it be that the download ID for that product is not valid any more? I don't see why it would be but perhaps when updating WooCommerce and applying the database changes that it missed that one product. Is there a way to reset the download ID for the product and apply to all orders that purchased it?

**Comment by seb86** [comment]
OK after some deeper looking into the post meta for my downloadable products, for some strange reason the **download_id** does not exist for the one product that I am having issues with.

This is how it is saved currently for the `_downloadable_files` meta.

Broken: `a:1:{s:47:"alnp-genesis-framework-support-child-theme-v1.1";a:3:{s:2:"id";s:47:"alnp-genesis-framework-support-child-theme-v1.1";s:4:"name";s:47:"alnp-genesis-framework-support-child-theme-v1.1";s:4:"file";s:23:"[amazon_s3 id="583736"]";}}`

And yet my other products are saved like so:

Working: `a:1:{s:32:"2a2d8b585a6201ff05cc29b85b02d9c4";a:2:{s:4:"name";s:20:"Product Name";s:4:"file";s:104:"https://domain.com/wp-content/uploads/woocommerce_uploads/2017/02/filename.zip";}}`

What would be the best approach to fix this?

I look forward to your response.

Thank you.

**Comment by mikejolley** [comment]
`alnp-genesis-framework-support-child-theme-v1.1` is too long. Are you sure a plugin isn't changing this ID?

Download ID is a 36 long hash, but 32 (legacy) work too.

```
a:1:{s:36:"b5104523-8738-4cb9-8c01-560a28a7baa4";a:3:{s:2:"id";s:36:"b5104523-8738-4cb9-8c01-560a28a7baa4";s:4:"name";s:6:"Single";s:4:"file";s:85:"https://demo.woothemes.com/woocommerce/wp-content/uploads/sites/56/2017/08/single.jpg";}}
```

Using filename is the issue.

**Comment by seb86** [comment]
> Are you sure a plugin isn't changing this ID?

I have shared my active plugins which have been the same for months and I don't think any of them would have changed the ID. At the moment all of my digital products are using the 32 legacy method.

I found the download ID of the product given to previous customers and manually updated the serialized data. It appears to be working now. Could we not protect the download meta to be **read-only** unless given strict permission to change it to prevent this type of issue in the future?

**Comment by mikejolley** [comment]
> I have shared my active plugins which have been the same for months and I don't think any of them would have changed the ID

Maybe they have been the same for months, but this ID structure is a more recent change.

**Comment by seb86** [comment]
Which version of WooCommerce was this ID structure change introduced?

**Comment by AndreiIgna** [comment]
Just came here to complain about this and see this is happening on other site as well 😃 

Can confirm there's a problem with this functionality. 

This product has 5 downloadable files
<img width="678" alt="screen shot 2018-06-07 at 15 16 42" src="https://user-images.githubusercontent.com/263021/41101975-f56abf80-6a65-11e8-8870-c1447b923d44.png">

But orders get only 4 downloadable files
<img width="907" alt="screen shot 2018-06-07 at 15 16 59" src="https://user-images.githubusercontent.com/263021/41101992-027cf7a6-6a66-11e8-9e67-3d40b971cc77.png">

When trying to manually add the product -> "Grant access", only 4 files are added, with no warning about the fifth. It's just ignored and didn't know why until now. 

Saw the character limit in database, but why are the files just discarded? Would be ok if the file name is trimmed or something similar is done

**Comment by mikejolley** [comment]
You're both using amazon files. What extension controls that?

**Comment by seb86** [comment]
WP Offload S3

**Comment by mikejolley** [comment]
@seb86 Can you disable that and test adding a file to a product- see if it gets correct hash. It's premium so I don't have access.

Hash changes came in 3.3.

**Comment by mikejolley** [comment]
Seems it must be; I'm unable to make WC save a filename. It only saves those hashes.

Happy to review it if you can email me a copy, otherwise you can confirm its that by disabling, saving a new downloadable file, and checking it has a has.

Closing for now as it doesn't appear to be a bug in core.

Thanks

**Comment by ianmjones** [comment]
Hi, lead developer for WP Offload S3 here.

With the help of @AndreiIgna ​I've found the problem and there will be a fix in the next release of WP Offload S3.

It was due to the new hidden hash input field inside the `file_name` table cell being accidentally set with the file name when an offloaded file was picked from the Media panel. This means any filename with over 32 characters in its name became the "download_id" when inserting into the `wp_woocommerce_downloadable_product_permissions` table. Unfortunately WooCommerce then silently fails to insert the record and the customer does not then get permission to download that file.
